File: Ruby/CivitasR/lib/Casilla.rb
#encoding:utf-8

require_relative "TituloPropiedad"
require_relative "Enum"
require_relative "Diario"
require_relative "Jugador"

module Civitas

  class Casilla_Descanso
    
    def initialize(nombre)
      @nombre = nombre
    end
    
    def informe(actual, todos)
      Diario.instance.ocurre_evento(
        "Casilla -- El jugador " + todos[actual].toString() +
        " ha caido en la casilla " + toString()
      )
    end
    protected :informe

    def recibeJugador(iactual,todos)
      informe(iactual,todos)
    end


    def toString()
      s = @nombre
      return s
    end
    
    def jugadorCorrecto(actual, todos)
      return actual < todos.size()
    end
    
    attr_reader :nombre
  end
  
  
  
  class Casilla_Sorpresa < Casilla_Descanso

    def initialize(mazo, nombre)
      super(nombre)
      @mazo = mazo
    end
    
    def recibeJugador(iactual,todos) 
      if(jugadorCorrecto(iactual,todos))
        sorpresa=@mazo.siguiente() #1
        informe(iactual,todos)
        sorpresa.aplicarAJugador(iactual,todos)
      end
    end

    def toString
      return @nombre + " ( Sorpresa )"
    end
  
  end
  
  class Casilla_Impuesto < Casilla_Descanso
    
    def initialize(cantidad, nombre)
     super(nombre)
      @cantidad = cantidad
    end
    
    def recibeJugador(actual, todos)
      if jugadorCorrecto(actual, todos)
        informe(actual, todos)
        actual[todos].pagaImpuesto(@importe)
      end
    end

    def toString
      return @nombre + " ( Impuesto, cantidad: " + @cantidad.to_str + " )"
    end
    
    attr_reader :cantidad
    
  end
  
  class Casilla_Juez < Casilla_Descanso
    @@carcel = 0
    
    def initialize(numCasillaCarcel, nombre)
      super(nombre)
      @@carcel = numCasillaCarcel
    end
    
    def recibeJugador_juez(actual, todos)
      if jugadorCorrecto(actual, todos)
        informe(actual, todos)
        todos[actual].encarcelar(@@carcel)
      end
    end

    def toString
      return @nombre + " ( juez )"
    end
    
    
  end
  
  
  
  class Casilla_Calle < Casilla_Descanso

    def initialize(titulo)
      super(titulo.nombre)
      @titulo = titulo
    end
    
    def recibeJugador(iactual,todos)
      if(jugadorCorrecto(iactual,todos))
        informe(iactual,todos)
        jugador=todos[iactual]
        if(!@titulo.tienePropietario)
          jugador.puedeComprarCasilla
        else
          @titulo.tramitarAlquiler(jugador)
        end
      end
    end
    
    def toString
      return @nombre + " ( Propiedad: " + @titulo.toString + " )"
    end
    
    attr_reader :titulo
    
  end
end

File: Java/Civitas/src/civitas/Casilla.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;

import java.util.ArrayList;

public class Casilla {
    private String nombre;
    private static int carcel;
    private float importe;
    private TipoCasilla tipo;
    private TituloPropiedad tituloPropiedad;
    private MazoSorpresas mazo;
    
    Casilla(String tnombre){
        tipo=TipoCasilla.DESCANSO;
        nombre=tnombre;
    }

    Casilla (TituloPropiedad titulo){
        init();
        tipo=TipoCasilla.CALLE;
        nombre=titulo.getNombre();
        tituloPropiedad=titulo;
    }

    Casilla (float cantidad,String tnombre){
        init();
        tipo=TipoCasilla.IMPUESTO;
        importe=cantidad;
        nombre=tnombre;
    }

    Casilla (int numCasillaCarcel,String tnombre){
        init();
        tipo=TipoCasilla.JUEZ;
        carcel=numCasillaCarcel;
        nombre=tnombre;
    }

    Casilla (MazoSorpresas tmazo,String tnombre){
        init();
        tipo=TipoCasilla.SORPRESA;
        mazo=tmazo;
        nombre=tnombre;
    }

    public String getNombre(){
        return nombre;
    }

    TituloPropiedad getTituloPropiedad(){
        return tituloPropiedad;
    }

    private void informe(int iactual,ArrayList<Jugador> todos){
        Diario.getInstance().ocurreEvento(todos.get(iactual).toString() + " " + toString());
    }

    private void init(){
		carcel=0;
		importe=0;
		tituloPropiedad=null;
		mazo=null;
    }

    public Boolean jugadorCorrecto(int iactual,ArrayList<Jugador> todos){
		return iactual<todos.size();
    }

    void recibeJugador(int iactual,ArrayList<Jugador> todos){
		switch(tipo){
			case CALLE:
				recibeJugador_calle(iactual,todos);
			case IMPUESTO:
				recibeJugador_impuesto(iactual,todos);
			case JUEZ:
				recibeJugador_juez(iactual,todos);
			case SORPRESA:
				recibeJugador_sorpresa(iactual,todos);
			default:
				informe(iactual,todos);
		}
    }

    private void recibeJugador_calle(int actual,ArrayList<Jugador> todos){
		if(jugadorCorrecto(actual, todos)){
			informe(actual,todos);
		}
		Jugador jugador = todos.get(actual);
		if( tituloPropiedad.tienePropietario() ){
			jugador.puedeComprarCasilla();
			tituloPropiedad.tramitarAlquiler(jugador);
		}
    }

    private void recibeJugador_impuesto(int actual,ArrayList <Jugador> todos){
        if(jugadorCorrecto(actual,todos)){
            this.informe(actual,todos);
            todos.get(actual).pagaImpuesto(importe);
        }
    }

    private void recibeJugador_juez(int actual,ArrayList <Jugador> todos){
        if(jugadorCorrecto(actual,todos)){
                informe(actual,todos);
                todos.get(actual).encarcelar(carcel);
        }
    }

    private void recibeJugador_sorpresa(int actual,ArrayList <Jugador> todos){
        if(jugadorCorrecto(actual,todos)){
			informe(actual, todos);
			mazo.siguiente().aplicarAJugador(actual, todos);
		}
    }

    public String toString(){
        return "Nombre: " + nombre;
    }

}
